core/shared/src/main/scala/scalaz/zio/IO.scala

```diff
@@ -753,7 +757,7 @@ sealed abstract class IO[+E, +A] extends Serializable { self =>
    * Keep or break a promise based on the result of this action.
    */
   final def to[E1 >: E, A1 >: A](p: Promise[E1, A1]): IO[Nothing, Boolean] =
-    self.run.flatMap(x => p.done(IO.done(x)))
+    self.redeem0(cause => p.done(IO.fail0(cause)), p.complete).onInterrupt(p.interrupt)
```

Comment:
With your above change, you can now leave this as:

```scala
 self.run.flatMap(x => p.done(IO.done(x)))
```

Bonus points: it'll be faster!

Reply:
Don't we need to leave .onInterrupt to still complete the promise in case of outer interruption?

Reply:
You're right! 👍